Optimize retrieving display name when searching for users in a group

CarlSchwan opened the pull request:
This is a recurrent scenario where we are searching for users and then for
each user, we fetch the displayName. This is inefficient, so instead try
to do one query to fetch everything (e.g. Database backend) or use the
already existing DisplayNameCache helper.

Splitted from https://github.com/nextcloud/server/pull/32201

CarlSchwan committed:
Optimize retrieving display name when searching for users in a group

This is recurrent scenario that we are searching for users and then for
each users we fetch the displayName. This is inefficient, so instead try
to do one query to fetch everything (e.g. Database backend) or use the
already existing DisplayNameCache helper.

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

come-nc commented on lib/private/User/Manager.php:
```diff
@@ -284,29 +284,27 @@
 	 */
 	public function search($pattern, $limit = null, $offset = null) {
 		$users = [];
+		$displayNameCache = \OCP\Sserver::get(DisplayNameCache::class);
```
:snake: :snake:

come-nc commented on lib/private/Group/Database.php:
```diff
@@ -377,6 +380,58 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
+
+		$users = [];
+		$userManager = \OC::$server->get(IUserManager::class);
+		$displayNameCache = \OC::$server->get(DisplayNameCache::class);
```
\OCP\Server?

come-nc commented on lib/private/Group/Group.php:
```diff
@@ -305,12 +300,12 @@ public function countDisabled() {
 	 * @param int $limit
 	 * @param int $offset
 	 * @return \OC\User\User[]
+	 * @depreacted 25.0.0 Use searchUsers instead (same implementation)
```
```suggestion
	 * @deprecated 25.0.0 Use searchUsers instead (same implementation)
```

come-nc commented on lib/public/GroupInterface.php:
```diff
@@ -107,13 +107,35 @@ public function getGroups($search = '', $limit = -1, $offset = 0);
 	 * @param int $offset
 	 * @return array an array of user ids
 	 * @since 4.5.0
+	 * @depreacted 25.0.0 Use searchInGroup instead, for performance reasons
```
```suggestion
	 * @deprecated 25.0.0 Use searchInGroup instead, for performance reasons
```

come-nc commented on tests/lib/Group/ManagerTest.php:
```diff
@@ -729,22 +731,9 @@ public function testDisplayNamesInGroupWithOneUserBackendAndSearchEmpty() {
 			->with('testgroup', '', -1, 0)
-			->willReturn(['user2', 'user33']);
-
-		$this->userManager->expects($this->any())
```
$this->never() ?

come-nc commented on lib/public/Group/Backend/ABackend.php:
```diff
@@ -65,4 +66,15 @@ public function implementsActions($actions): bool {
+
+	public function searchInGroup(string $gid, string $search = '', int $limit = -1, int $offset = 0): array {
+		// Default implementation for compatibility reasons
+		$displayNameCache = Server::get(DisplayNameCache::class);
```
missing use statement for DisplayNameCache, IUserManager and LazyUser

icewind1991 commented on lib/private/Group/Database.php:
```diff
@@ -377,6 +380,55 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
 		return $users;
 	}
 
+	public function searchInGroup(string $gid, string $search = '', int $limit = -1, int $offset = 0): array {
```
Given that this seems to do the same amount of work as `usersInGroup`, it makes sense to have `usersInGroup` call this and map the results to just the uid

icewind1991 commented on lib/private/Group/Database.php:
```diff
@@ -377,6 +380,55 @@ public function usersInGroup($gid, $search = '', $limit = -1, $offset = 0) {
+			->where($query->expr()->eq('gid', $query->createNamedParameter($gid)))
+			->orderBy('g.uid', 'ASC');
+
+		$query->leftJoin('g', 'users', 'u', $query->expr()->eq('g.uid', 'u.uid'))
```
missing `;`

icewind1991 commented on lib/public/GroupInterface.php:
```diff
@@ -107,13 +107,35 @@ public function getGroups($search = '', $limit = -1, $offset = 0);
+	 * @return IUser[]
+	 * @since 25.0.0
+	 */
+	public function searchInGroup(string $gid, string $search = '', int $limit = -1, int $offset = 0): array;
```
for compatibility with 3rdparty group backends it might make more sense to add a new extension interface for this and have the group manager handle the fallback to `usersInGroup`+`LazyUser`

CarlSchwan replied:
Done

CarlSchwan committed:
Split new method in a new group backend interface

Better for backward compatibility, also move new interfaces to nc 26

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

come-nc commented on lib/private/User/Manager.php:
```diff
@@ -291,58 +291,52 @@ public function checkPasswordNoLogging($loginName, $password) {
 	 */
 	public function searchDisplayName($pattern, $limit = null, $offset = null) {
 		$users = [];
+		$displayNameCache = \OCP\Server::get(DisplayNameCache::class);
```
unused?

come-nc committed:
Make code clearer and bump @ deprecated annotations

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Let OC\Group\Group handle the fallback and remove default implementation from ABackend

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
We should decide if the array returned by usersInGroup and searchInGroup is using the uid as key or integers.
This is currently not consistent and why the tests fails.

I would say usersInGroup need to use int for BC, but searchInGroup could use uid as it may be handy in some cases.

come-nc committed:
Clear up return types

usersInGroup index by int for BC, searchInGroup index by uid (string).

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
This is ready for review I think.
I would like to test and confirm that it reduces requests though.

come-nc committed:
Update version number in since and deprecated annotations

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Use the searchDisplayName recommended method in user:list

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix tests, and fix Group::searchUsers to avoid duplicates

Also went back to searchUsers indexing the array by uid as it was the
 previous behavior and the IGroup phpdoc does not say anything about the
 keys.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>
